### PR TITLE
[MTDSA 12078] Controller for Create a Historic Non-FHL UK Property Income & Expenses Period Summary

### PR DESCRIPTION
**Pull request by sean-flynn-HMRC**
Controller & ITs for Create a Historic Non-FHL UK Property Income & Expenses Period Summary

### CODE REVIEW

File: it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+      |           "rentsReceived": 567.56
+      |         }
+      |        },
+      |       "expenses":{
+      |          "consolidatedExpenses": 235.78
+      |     }
+      |}""".stripMargin
+  )
+
+  val allInvalidValueRequestError: MtdError = ValueFormatError.copy(
```
Comment (jeremystone) [inline it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala:237]: You could just have a single bad value to prove that `ValueFormatError` is picked up. (Testing that we pick up multiple errors is more at home in a validation unit test than an integration test - and I suspect we already do this.) Then you can move this to an in-line value at the point where it is used.

File: it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+import play.api.http.HeaderNames.ACCEPT
+import play.api.http.Status
+import play.api.libs.json.{JsValue, Json}
+import play.api.libs.ws.{WSRequest, WSResponse}
+import play.api.test.Helpers.AUTHORIZATION
+import support.V2IntegrationBaseSpec
+import v2.models.errors._
+import v2.stubs.{AuditStub, AuthStub, DownstreamStub, MtdIdLookupStub}
+
+class CreateHistoricNonFHlUkPiePeriodSummaryControllerISpec extends V2IntegrationBaseSpec {
```
Comment (jeremystone) [inline it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala:29]: Class name doesn't line up with file name:
```suggestion
class CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec extends V2IntegrationBaseSpec {
```

File: it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+
+        val input = Seq(
+          ("AA1123A", requestBodyJson, Status.BAD_REQUEST, NinoFormatError),
+          ("AA123456A", invalidToDateRequestJson, Status.BAD_REQUEST, ToDateFormatError),
+          ("AA123456A", invalidFromDateRequestJson, Status.BAD_REQUEST, FromDateFormatError),
+          ("AA123456A", bothExpensesSuppliedRequestJson, Status.BAD_REQUEST, RuleBothExpensesSuppliedError.copy(paths=Some(Seq("/expenses/consolidatedExpenses")))),
+          ("AA123456A", missingFromAndToDateRequestJson, Status.BAD_REQUEST,  RuleIncorrectOrEmptyBodyError.copy(paths=Some(Seq("/fromDate", "/toDate")))),
+          ("AA123456A", toDateBeforeFromDateRequestJson, Status.BAD_REQUEST, RuleToDateBeforeFromDateError),
+          ("AA123456A", invalidValueRequestJson, Status.BAD_REQUEST, allInvalidValueRequestError)
+        ) //Should we add a test for rentARoom.amountClaimed and rentARoom.rentsReceived being missing? Tech spec says they're mandatory, business spec says they're optional.
```
Comment (jeremystone) [inline it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala:390]: Code has them as optional - so presumably in agreement with RAML. 

I believe the backend values are optional and there was a question over whether we should make them mandatory, but it looks like we've done what we did with the similar non-historic endpoints. So I suspect the tech spec mappings need adjusting.

File: app/v2/controllers/CreateHistoricNonFHLUkPiePeriodSummaryController.scala [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+      logger.info(message = s"[${endpointLogContext.controllerName}][${endpointLogContext.endpointName}] with correlationId : $correlationId")
+
+      val rawData = CreateHistoricNonFhlUkPropertyPeriodSummaryRawData(nino, request.body)
+      val result =
+        for {
+          parsedRequest   <- EitherT.fromEither[Future](parser.parseRequest(rawData))
+          serviceResponse <- EitherT(service.createPeriodSummary(parsedRequest))
+          vendorResponse <- EitherT.fromEither[Future](
+            hateoasFactory
+              .wrap(serviceResponse.responseData, CreateHistoricNonFhlUkPiePeriodSummaryHateoasData(nino, serviceResponse.responseData.periodId.get, serviceResponse.responseData.transactionReference))
```
Comment (jeremystone) [inline app/v2/controllers/CreateHistoricNonFHLUkPiePeriodSummaryController.scala:66]: Best not to use unprotected `get` on an `Option`. Form the periodId from the request body instead.

File: test/v2/controllers/CreateHistoricNonFhlUkPiePeriodSummaryControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+            contentAsJson(result) shouldBe Json.toJson(mtdError)
+            status(result) shouldBe expectedStatus
+            header("X-CorrelationId", result) shouldBe Some(correlationId)
+          }
+        }
+
+        val input = Seq(
+          (NinoFormatError, BAD_REQUEST),
+          (NotFoundError, NOT_FOUND),
+          (RuleDuplicateSubmissionError, BAD_REQUEST),
```
[the remark this answers is not in the document]
Comment (sean-flynn-HMRC) [reply test/v2/controllers/CreateHistoricNonFhlUkPiePeriodSummaryControllerSpec.scala:184]: Added 
image

File: app/v2/controllers/requestParsers/validators/CreateHistoricNonFhlUkPropertyPeriodSummaryValidator.scala [hunk 6]
```diff
@@ -50,22 +47,32 @@ class CreateHistoricNonFhlUkPropertyPeriodSummaryValidator @Inject()(appConfig:
...
+
   private def validateBody(body: CreateHistoricNonFhlUkPropertyPeriodSummaryRequestBody): Either[List[MtdError], Unit] = {
 
     val formatDateErrors = DateValidation.validate(body.fromDate, true) ++
       DateValidation.validate(body.toDate, false)
 
-    val historicTaxPeriodYearErrors = HistoricTaxPeriodYearValidation.validate(minTaxYear, maxTaxYear, body.fromDate) ++
-      HistoricTaxPeriodYearValidation.validate(minTaxYear, maxTaxYear, body.toDate)
+    def validateToDateIsAfterFromDate(toDate: String, fromDate:String):List[MtdError] ={
+      if(DateValidation.validate(fromDate, true) == NoValidationErrors && DateValidation.validate(toDate, false) == NoValidationErrors){
```
[the remark this answers is not in the document]
Comment (sean-flynn-HMRC) [reply app/v2/controllers/requestParsers/validators/CreateHistoricNonFhlUkPropertyPeriodSummaryValidator.scala]: Updated

File: test/v2/controllers/CreateHistoricNonFhlUkPiePeriodSummaryControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+      idGenerator = mockIdGenerator
+    )
+
+    MockMtdIdLookupService.lookup(nino).returns(Future.successful(Right("test-mtd-id")))
+    MockedEnrolmentsAuthService.authoriseUser()
+    MockIdGenerator.getCorrelationId.returns(correlationId)
+  }
+
+  private val requestBody =
+    CreateHistoricNonFhlUkPropertyPeriodSummaryRequestBody(fromDate = "2019-03-11", toDate = "2020-04-23", Some(incomeModel), Some(expensesModel))
```
[the remark this answers is not in the document]
Comment (sean-flynn-HMRC) [reply test/v2/controllers/CreateHistoricNonFhlUkPiePeriodSummaryControllerSpec.scala]: Updated

File: test/v2/controllers/CreateHistoricNonFhlUkPiePeriodSummaryControllerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+      |""".stripMargin
+  )
+
+  private val requestData =
+    CreateHistoricNonFhlUkPropertyPeriodSummaryRequest(nino = Nino(nino), body = requestBody)
+  private val rawData = CreateHistoricNonFhlUkPropertyPeriodSummaryRawData(nino = nino, body = requestBodyJson)
+
+  private val testHateoasLinks =
+    Seq(
+      Link(
```
[the remark this answers is not in the document]
Comment (sean-flynn-HMRC) [reply test/v2/controllers/CreateHistoricNonFhlUkPiePeriodSummaryControllerSpec.scala]: Updated

File: it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala (hunk 1 above)
Comment (sean-flynn-HMRC) [reply it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala:237]: Simplified + updated

File: it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala (hunk 2 above)
Comment (sean-flynn-HMRC) [reply it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala]: Updated

File: it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala (hunk 3 above)
Comment (sean-flynn-HMRC) [reply it/v2/endpoints/CreateHistoricNonFhlUkPiePeriodSummaryControllerISpec.scala]: Removed the comment

File: app/v2/controllers/CreateHistoricNonFHLUkPiePeriodSummaryController.scala (hunk 4 above)
Comment (sean-flynn-HMRC) [reply app/v2/controllers/CreateHistoricNonFHLUkPiePeriodSummaryController.scala]: Updated

**Pull request merged**